sql/src/main/java/io/crate/expression/reference/doc/lucene/LuceneReferenceResolver.java

```diff
@@ -63,61 +65,58 @@ public LuceneReferenceResolver(FieldTypeLookup fieldTypeLookup, IndexSettings in
     }
 
     @Override
-    public LuceneCollectorExpression<?> getImplementation(Reference refInfo) {
-        assert refInfo.granularity() == RowGranularity.DOC : "lucene collector expressions are required to be on DOC granularity";
-
-        final DataType dataType = refInfo.valueType();
-        DataType innerType = CollectionType.unnest(dataType);
-        if (innerType.id() == ObjectType.ID || innerType.id() == GeoShapeType.ID) {
-            // FetchOrEval takes care of rewriting the reference to do a source lookup.
-            // However, when fetching is disabled, this does not happen and we need to ensure
-            // (at least) for objects (also if inside arrays) that we do a source lookup.
-            refInfo = DocReferences.toSourceLookup(refInfo);
-        }
+    public LuceneCollectorExpression<?> getImplementation(Reference ref) {
+        assert ref.granularity() == RowGranularity.DOC : "lucene collector expressions are required to be on DOC granularity";
+
+        ColumnIdent column = ref.column();
+        switch (column.name()) {
+            case DocSysColumns.Names.RAW:
+                if (column.isTopLevel()) {
+                    return new RawCollectorExpression();
+                }
+                throw new UnsupportedFeatureException("_raw expression does not support subscripts: " + column);
+
+            case DocSysColumns.Names.UID:
+                if (indexSettings.isSingleType()) {
+                    return new IdCollectorExpression(); // _uid == _id
+                }
+                return new UidCollectorExpression();
+
+            case DocSysColumns.Names.ID:
+                if (indexSettings.isSingleType()) {
+                    return new IdCollectorExpression();
+                }
+                return new IdFromUidCollectorExpression();
+
+            case DocSysColumns.Names.DOC:
+                return DocCollectorExpression.create(ref);
 
-        ColumnIdent columnIdent = refInfo.column();
-        String name = columnIdent.name();
-        if (RawCollectorExpression.COLUMN_NAME.equals(name)) {
-            if (columnIdent.isTopLevel()) {
-                return new RawCollectorExpression();
-            } else {
-                // TODO: implement an Object source expression which may support subscripts
```

Comment:
Is it ok to remove this TODO?

Reply:
Yep, It doesn't add more information other than that it's not yet implemented and has been there for ages.